### PR TITLE
ci: GH workflow for publishing docker images

### PR DESCRIPTION
**Pull request by bermuell**
## Description

Closes: #1592 
GH workflow to publish ICS docker images

Workflow is triggerd to build
- `interchain-security:latest` image when a commit is pushed to 'main' branch
- `interchain-security:vX.Y.Z` when a tag with pattern vX.Y.Z is pushed 
- `interchain-security:vX.Y.Z-lsm` when a tag with pattern vX.Y.Z-lsm is pushed 

for all other commits/tags no docker image workflow  will be triggered

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Provided a link to the relevant issue or specification
- [ ] Reviewed "Files changed" and left comments if necessary 
- [ ] Confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] Confirmed all author checklist items have been addressed
- [ ] Confirmed that this PR does not change production code

### CODE REVIEW

**Review by p-offtermatt — COMMENTED** [review]
Looks reasonable to me.
Unfortunately, iirc the only realistic way to test this is to merge and
check what it publishes. An alternative might be forking the repo to informalsystems and playing with it there (at least we don't mess with the real repo)

wdyt?

**Comment by bermuell** [comment]
> Looks reasonable to me. Unfortunately, iirc the only realistic way to test this is to merge and check what it publishes. An alternative might be forking the repo to informalsystems and playing with it there (at least we don't mess with the real repo)
> 
> wdyt?

yes you're right. did not find a way to test this outside the PR. will observe what is published and adapt if needed.

**Review by p-offtermatt — APPROVED** [review]
Workflow looks good as far as I can tell, so approval and let's see whether it works as expected.

**Review by MSalopek — COMMENTED** [review]
Added some comments.

Let's reduce the amount of images being pushed. The most important events happen on push to main and when a tag/release is created.

File: .github/workflows/docker-publish.yml [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+# This workflow uses actions that are not certified by GitHub.
+# They are provided by a third-party and are governed by
+# separate terms of service, privacy policy, and support
+# documentation.
+
+on:
+  release:
+    types: [published]
+  push:
+    branches: [ "main", "release/v4.0.x", "release/v3.3.x", "release/v3.2.x", "release/v3.1.x", "release/v3.0.x"]
```
[the remark this answers is not in the document]
Comment (p-offtermatt) [reply .github/workflows/docker-publish.yml:16]: Keep in mind we want this to build the images for compatibility tests. We should include some way of pushing the images for releases

[the remark this answers is not in the document]
Comment (p-offtermatt) [reply .github/workflows/docker-publish.yml:16]: Ah, I guess part of this is taken care of by the "published releases" above, so I think it's probably ok to remove the releases here

**Pull request merged**